provider/module/module.go

```diff
@@ -0,0 +1,65 @@
+package module
+
+import (
+	"embed"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/hashicorp/go-hclog"
+)
+
+// InfoReader is called when the user executes a module, to get provider supported metadata about the given module
+type InfoReader func(logger hclog.Logger, module string, prefferedVersions []uint32) (resp InfoResponse, err error)
+
+// InfoResponse is what the provider returns from an InfoReader request.
+type InfoResponse struct {
+	// Version of the supplied "info"
+	Version uint32
+	// Info, in the given version
+	Info map[string][]byte
+	// All versions supported by the provider, if any
+	SupportedVersions []uint32
+}
+
+// Serve returns an InfoReader handler given a "moduleData" filesystem.
+// The fs should have all the required files for the modules in a moduledata/ directory, as one subdirectory per module ID.
+// Filenames are the version IDs with an .hcl extension added.
+func Serve(moduleData embed.FS) InfoReader {
+	return func(logger hclog.Logger, module string, prefferedVersions []uint32) (resp InfoResponse, err error) {
```

Comment:
I am not a fan of this implementation, and hiding the response var initialization the return value.

For one I would read the embdded module info before returning the function creating a mapping of module -> versions.  Would make the check easier in the function i.e only check if exists.

If it does exists read all files (also recursively in directory) returning all files as the version configuration files.

Reply:
I don't think we could do recursively. Maybe in next version. swapped the order.